### PR TITLE
Fix race condition in epub progress saving during scroll events

### LINKED ISSUE #4446: EPUB Reader: Progress syncing is off when switching between one- and two-column mode
### What happened?

When switching from a device with one-column mode to a device with two-column mode the virtual page might not be computed correctly according to the reading progress and jumps back to the top of the native page (as per EPUB metadata).
I have attached a video, where I have bookmarked the top of the page and when switching to another instance with a different column mode it jumps back to the bookmark. Happens to me regularly when I read on my smartphone and then switch to my tablet or desktop device with two-column mode.

https://github.com/user-attachments/assets/145d7a4e-5cde-41df-bc0f-c7d50193708b

### What did you expect?

The EPUB reader will start roughly at the position where last closed, unaffected by the reading mode. (During scroll mode and from one-column to one-column I didn't face any issues)

### Kavita Version Number - If you don't see your version number listed, please update Kavita and see if your issue still persists.

0.8.9.1 - Stable

### Are you accessing kavita through a reverse proxy? If yes, confirm that the issue persists with a direct connection

Yes

### What operating system is Kavita being hosted from?

Docker (Dockerhub Container)

### If the issue is being seen on Desktop, what OS are you running where you see the issue?

Windows

### If the issue is being seen in the UI, what browsers are you seeing the problem on?

Chrome

### If the issue is being seen on Mobile, what OS are you running where you see the issue?

iOS

### If the issue is being seen on the Mobile UI, what browsers are you seeing the problem on?

Safari

### Relevant log output

```shell

```

### Additional Notes

_No response_

### PR DESCRIPTION
**Pull request by vipin2310**
# Fixed
- Fixed: If a scroll event is pending (e.g. initial load), the reader now ignores scroll events and doesn't update/save the progress. This caused a race condition where reopening a book would often reset the reading position and progress to the top of the page (often chapter), due to a timing issue during page load. (Fixes: #4446)

### CODE REVIEW

**Comment by majora2007** [comment]
Appreciate the pr, I'm on holiday, but I will take a look at this in a week or so. @Fesaa if you want to give this a shake out, please do.

**Comment by vipin2310** [comment]
This was already catched in the `applyPageStyles` function:
```C#
// When the user switches pages, there may be a pending scroll that moves to the start or end of the page
// For example, `this.scrollWithinPage(...)` might be triggered when the user presses the prev/next page button
// So, we don't need to do another page scroll here
if (!this.hasDelayedScroll) {
  this.scrollTo(resumeElement);
}
```

So I also added the bypass to `handleScrollEvents` and `snapScrollOnResize`.

**Comment by Fesaa** [comment]
This doesn't work, there's no progress saved after you go to a next page (until you page again). I.e. if you page to a next page (/chapter) then quit the reader, you'll open it again in the previous one (As opposed to the last virtual page from before).

https://github.com/user-attachments/assets/bb858ed3-2c24-487f-8af5-4443982e6e84

We'll need to account for the scrolling happening, and delay the progress event instead of skipping it.

**Comment by vipin2310** [comment]
Thanks for catching that, I haven't tested that scenario. I will try to take a look at this today or tomorrow.

**Comment by vipin2310** [comment]
I have now reworked this PR according to your suggestion @Fesaa, this also fixed the progress issue that I introduced before (see your video).
I have changed the location, where the progress event will be called (after scroll finishes) and also instead of skipping it gets delayed if there is a pending scroll. Progress saving should hopefully be more reliable now.

However, now there are more nested callbacks (afterFrame -> setTimeout -> afterFrame):
```typescript
    // `afterFrame() + setTimeout()` can likely be replaced with `requestAnimationFrame()` instead
    afterFrame(() => {
      setTimeout(() => {
        this.hasDelayedScroll = false;
        lambda();
        afterFrame(() => this.handleScrollEvent());
      }, SCROLL_DELAY)
    });
```

I could replace them with async/awaits by wrapping them in Promises. It could make it more readable, but I don't know If you would prefer that pattern.

**Comment by vipin2310** [comment]
After testing more thoroughly I would like to summarize my findings before and after the changes in this PR. For the ease of explaining, I would refer to the native pages in the epubs as "chapter" and to the virtual pages in column mode as "pages.

**Before:**
1. If you open a book in column-mode, don't page again, and exit (or switch column-mode: Progress will be resetted to the top of the chapter.

  https://github.com/user-attachments/assets/b53d2c11-b956-44b4-9aa9-9b2830e21d44

2. If you page to a new chapter in column-mode, don't page again, and exit: Progress will be set to the end of the chapter, if you have already been at the end of previous chapter before paging.

  https://github.com/user-attachments/assets/3a512347-61b2-429f-9551-61ba77435254


3. If you are in scroll mode and page to a new chapter, and immediately exit before scrolling to the top finished: Progress will be set to the end of the chapter. This might be related to (2)

  https://github.com/user-attachments/assets/d0d220f5-7a2e-4d8b-8b12-5e1b1d211e95


**After**:
(1) and (2) are fixed.
(3) still persists because the progress event wasn't called before closing. But this might be an edge-case as you would need to exit faster than the scrolling finished. In the future one could maybe work on bypassing the scroll to the top when paging to a new chapter? I haven't looked into this yet, since it's not directly related to the issue I am trying to solve.

If you want I can have a look into (3) in another PR at some time.

**Comment by Fesaa** [comment]
Hello, thanks for the updates. You'll need the merge develop has some merge conflicts have popped up. 

Issue 3) is fine to keep as an edge case. I'll pull down and give it another test once develop has been merged in. But looks good from looking at the commit

File: UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts [hunk 1]
```diff
@@ -761,6 +761,12 @@ export class BookReaderComponent implements OnInit, AfterViewInit, OnDestroy {
    * Updates the TOC current page anchor, last scene path and saves progress
    */
   handleScrollEvent(bypassSave: boolean = false) {
+    // If a scroll is pending (e.g. initial load), do not update/save progress based on the intermediate scroll position
+    // (e.g. top of page). Delay the progress event until the pending scroll finished.
+    if (this.hasDelayedScroll) {
+      setTimeout(() => this.handleScrollEvent(bypassSave), SCROLL_DELAY);
+      return;
```
[the remark this answers is not in the document]
Comment (vipin2310) [reply UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:773]: Okay, thats valid. I can work on that later today

[the remark this answers is not in the document]
Comment (vipin2310) [reply UI/Web/src/app/book-reader/_components/book-reader/book-reader.component.ts:773]: Fixed in de36de5

**Comment by vipin2310** [comment]
I reference issue #3022 here as well, the last few comments seem to be related to the issues I presented in this comment: https://github.com/Kareadita/Kavita/pull/4452#issuecomment-3979759225 .

**Review by Fesaa — APPROVED** [review]

**Comment by Fesaa** [comment]
Thank you for the contribution, we'll release this with the next nightly! ❤️
